src/arithmetic/string_funcs/string_funcs.c

```diff
@@ -160,9 +182,20 @@ SIValue AR_SUBSTRING(SIValue *argv, int argc, void *private_data) {
 		length = MIN(length, suffix_len);
 	}
 
-	char *substring = rm_malloc((length + 1) * sizeof(char));
-	strncpy(substring, original + start, length);
-	substring[length] = '\0';
+	utf8proc_int32_t c;
+	int64_t start_bytes = 0;
+	for (int i = 0; i < start; i++) {
```

Comment:
for each one of the newly added loops, please add a comment explaining the purpose of the loop.